opm/simulators/flow/TracerModel.hpp

```diff
@@ -489,39 +704,104 @@ class TracerModel : public GenericTracerModel<GetPropType<TypeTag, Properties::G
             // Note that we solve for a concentration update (compared to previous time step)
             // Confer also assembleTracerEquations_(...) above.
             std::vector<TracerVector> dx(tr.concentration_);
-            for (int tIdx = 0; tIdx < tr.numTracer(); ++tIdx)
+            for (int tIdx = 0; tIdx < tr.numTracer(); ++tIdx) {
                 dx[tIdx] = 0.0;
+            }
 
             bool converged = this->linearSolveBatchwise_(*tr.mat, dx, tr.residual_);
             if (!converged) {
                 OpmLog::warning("### Tracer model: Linear solver did not converge. ###");
             }
 
             for (int tIdx = 0; tIdx < tr.numTracer(); ++tIdx) {
-                tr.concentration_[tIdx] -= dx[tIdx];
-                // Tracer concentrations for restart report
-                this->tracerConcentration_[tr.idx_[tIdx]] = tr.concentration_[tIdx];
+                for (std::size_t globalDofIdx = 0; globalDofIdx < tr.concentration_[tIdx].size(); ++globalDofIdx) {
+                    // New concetration. Concentrations that are negative or where free/solution phase is not
+                    // present are set to zero
+                    const auto& intQuants = simulator_.model().intensiveQuantities(globalDofIdx, 0);
+                    const auto& fs = intQuants.fluidState();
+                    Scalar Sf = decay<Scalar>(fs.saturation(tr.phaseIdx_));
+                    Scalar Ss = 0.0;
+                    if (tr.phaseIdx_ == FluidSystem::gasPhaseIdx && FluidSystem::enableDissolvedGas())
+                        Ss = decay<Scalar>(fs.saturation(FluidSystem::oilPhaseIdx));
+                    else if (tr.phaseIdx_ == FluidSystem::oilPhaseIdx && FluidSystem::enableVaporizedOil())
+                        Ss = decay<Scalar>(fs.saturation(FluidSystem::gasPhaseIdx));
+
+                    if ((tr.concentration_[tIdx][globalDofIdx][0] - dx[tIdx][globalDofIdx][0] < 0.0) 
+                        || Sf < 1e-6)
+                        tr.concentration_[tIdx][globalDofIdx][0] = 0.0;
```

Comment:
I think it would be better to use Scalar tol_gas_sat = 1e-6 to make it clear what it is.

Reply:
Changed.